### DISCUSSION TITLE
Vulkan Instance Creation Fails with VK_KHR_surface Ext. on Old Hardware

### DISCUSSION THREAD

**Discussion by TheVave** [dotnet/Silk.NET#2241] [Q&A] [answered]
I am switching from OpenGL to Vulkan and, while adding extensions, it suddenly never was able to create an instance. 

I eventually found that with VK_KHR_surface the result from CreateInstance was ErrorExtentionNotPresent. After a bit of research I thought this meant I need to update my drivers (and hardware from a GTX 980 GPU which is 10 years old). 

Though, I know that a game I play some, Valheim, works with the "Play with Vulkan" option, so I thought this was an error with my code.

### Code for instance creation ###

```
Vk vk = Vk.GetApi();

ApplicationInfo appInfo = new()
{
	SType = StructureType.ApplicationInfo,
	PApplicationName = (byte*)Marshal.StringToHGlobalAnsi(title),
	ApplicationVersion = new Version32(1, 0, 0),
	PEngineName = (byte*)Marshal.StringToHGlobalAnsi("No Engine"),
	EngineVersion = new Version32(1, 0, 0),
	ApiVersion = Vk.Version12
};

InstanceCreateInfo createInfo = new()
{
	SType = StructureType.InstanceCreateInfo,
	
	PApplicationInfo = &appInfo
};

uint c = 0;

string[] ext = ["VK_KHR_surface"];
c = (uint)ext.Length;

// just IntPtr.Zero as a byte*
byte* firstByte = Utils.NULLBYTEPTR;
fixed (string* firstExtPtr = &ext[0])
{
	firstByte = (byte*)firstExtPtr;
}

byte** bpp = &firstByte;
createInfo.EnabledExtensionCount = c;
createInfo.PpEnabledExtensionNames = bpp;

createInfo.EnabledLayerCount = 0;
Silk.NET.Vulkan.Result res = vk.CreateInstance(createInfo, null, out inst);
if (res != Silk.NET.Vulkan.Result.Success)
{
        // error handling code that makes a message box
	ErrorHandler.ThrowError(26, [res.ToString()], true);
}

Marshal.FreeHGlobal((IntPtr)appInfo.PApplicationName);
Marshal.FreeHGlobal((IntPtr)appInfo.PEngineName);
SilkMarshal.Free((nint)createInfo.PpEnabledExtensionNames);

if (EnableValidationLayers)
{
	SilkMarshal.Free((nint)createInfo.PpEnabledLayerNames);
}
```

What should I do?
Thanks for any help!

PS: Sorry if my code is messy, I'm following a C++ tutorial for Vulkan and translating it to C#.

**Comment by Perksey** [comment, marked as answer]
```cs
fixed (string* firstExtPtr = &ext[0])
```
This is not correct, this is taking a pointer to managed object reference (your compiler should warn you about this)

We have a convenience method `SilkMarshal.StringArrayToPtr` you can use to get a `byte**` from a `string[]`.

**Reply by TheVave** [reply]
So should I change it to this?
```
string[] ext = ["VK_KHR_surface"];
count = (uint)ext.Length;

byte* bp = (byte*)SilkMarshal.StringArrayToPtr(ext);
byte** bpp = &bp;
```

**Reply by Perksey** [reply]
Not quite: `(byte**)SilkMarshal.StringArrayToPtr(ext)`

Don't forget to free it after you've used it!

**Reply by TheVave** [reply]
Thanks! I didn't think that the fixed expression was the culprit. Could you say why that was the issue?

**Reply by Perksey** [reply]
`string` is a managed object and is itself a pointer to its internal data, which you can’t [easily] access. You have to either get a reference to its data, which GetPinnableReference does (but in UTF-16 format) or manually convert it to some raw memory.

**Reply by TheVave** [reply]
Thanks!
